### PR TITLE
WP-34, 35, 36, 40, 50: Complete FE for CC Page, Add Exercise to Workout

### PR DESCRIPTION
**Pull request by ltan02**
**Description:**
Split the Creator Corner's page into three different sections: `workout_viewer`, `exercise_library`, and `workout_editor`. For the `workout_viewer`, I used the same implementation that @AndiXiong used for her page (which uses the `ChewieListItem` and the `VideoPlayerController`. Similarly for the `exercise_library`, I used @aleccumming's implementation of `GridView` to create the exercise cards. The `workout_editor` acts as a `DragTarget` and allows any exercise card to be dragged onto it and be added to the workout. 

_Notes:_ 
- Currently, only the All and Saved buttons work.
- A `SliverGeometry is not valid: The "scrollExtent" is negative` exception is being thrown whenever you run the program and that is because the `chosenExercises` from the `workout_editor` is empty at the start, the exception itself does not affect the program.
- When running the code, please use: `flutter run -d chrome --web-renderer html` in the command line.
- A couple of the stylings are currently hard-coded, so there may be some pixel overflows when running on your own computers.

**Screenshots of Creator's Corner Page:**
image

image

**Changes:**
- Created a `styles.dart` file in the `lib/constants/` folder to store all the TextStyle. 
- Added the fonts that @ElinaJiang put in the Hi-Fi design (Baloo and Baloo Bhai 2). 
- Edited the design of the navigation bar by changing colour to our theme and including an area for the logo and area for the user’s profile picture. 
- Implemented the FE for the `workout_viewer`, `exercise_library`, and `workout_editor`.
- Implemented `Draggable` and `DragTarget` to add exercises to a workout.
- Using the Exercise model that @AndiXiong made, I added the `tags` property, which is of type `List<String>`.

**Related JIRA Tickets:**
https://ubclaunchpad.atlassian.net/browse/WP-34
https://ubclaunchpad.atlassian.net/browse/WP-35
https://ubclaunchpad.atlassian.net/browse/WP-36
https://ubclaunchpad.atlassian.net/browse/WP-40
https://ubclaunchpad.atlassian.net/browse/WP-50

### CODE REVIEW

File: lib/models/pair.dart [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class Pair<T1, T2> {
```
Comment (richardyjtian) [inline lib/models/pair.dart]: Do we need our own pair class? We could probably just use a final list of dynamic elements with size 2

File: lib/widgets/creators_corner/workout_editor.dart [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        body,
+        style: workoutEditorStyle,
+      ),
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      color: const Color(0xFFFFA101),
```
Comment (richardyjtian) [inline lib/widgets/creators_corner/workout_editor.dart]: We should reference this color and fonts below from the styles.dart class. I made this change in my local branch already, so we can maybe just open a ticket for this, and I'll make the changes after I make a pr for my local branch.

File: lib/widgets/navigation_bar/custom_tab_bar.dart [hunk 3]
```diff
@@ -14,23 +14,70 @@ class CustomTabBar extends StatelessWidget {
...
+                shape: BoxShape.circle,
+                color: Color(0xFF9D9494),
+              ),
+            ),
+          ),
+        ],
       ),
     );
+    
+    // return Padding(
```
Comment (richardyjtian) [inline lib/widgets/navigation_bar/custom_tab_bar.dart]: Can we remove this?

**Review by richardyjtian — COMMENTED** [review]
Could you open a ticket to fix the overflow errors later on and link it in the classes of the overflowing widgets?
Capture
Other than that, mostly everything looks great!

**Review by aleccumming — DISMISSED** [review]
All looks pretty good to me! Just the little comments @richardyjtian left.

File: lib/models/exercise.dart [hunk 4]
```diff
@@ -0,0 +1,33 @@
+// ignore_for_file: file_names
+
+class Exercise {
+  final int id;
+  final String name;
+  final String description;
+  final String thumbnailSrc;
+  final String videoSrc;
+  final int length; // units?
```
Comment (lilyydu) [inline lib/models/exercise.dart]: Update this to be in seconds!

File: lib/ui/creators_corner_page.dart [hunk 5]
```diff
@@ -1,16 +1,43 @@
...
-    return Center(
-      child: Container(
-        child: const Text(
-          "Creator's Corner Page",
-          style: TextStyle(fontSize: 30),
+    return Column(
+      crossAxisAlignment: CrossAxisAlignment.stretch,
+      children: <Widget>[
+        Expanded(
+          flex: 6,
```
Comment (lilyydu) [inline lib/ui/creators_corner_page.dart]: Small code smell but probably better to define these numbers as constants at the top

File: lib/widgets/creators_corner/exercise_library.dart [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+  bool sideBarOpened = false;
+
+  @override
+  void initState() {
+    super.initState();
+    String dummyThumbnailUrl = "https://cdn.centr.com/content/17000/16775/images/landscapewidemobile3x-bobby-push-up-16-9.jpg";
+    String dummyVideoUrl = "https://flutter.github.io/assets-for-api-docs/assets/videos/bee.mp4";
+    allExercises = [
+      Exercise(1, "Crunch", "Exercise1 Description", "https://teamworkoutplatform.s3.us-west-2.amazonaws.com/Exercise+1+Thumbnail.png", "https://teamworkoutplatform.s3.us-west-2.amazonaws.com/Exercise+1+.mp4", 1, ["abs", "legs"]),
```
Comment (lilyydu) [inline lib/widgets/creators_corner/exercise_library.dart]: I've added the dummy data into the Slack!

File: lib/widgets/creators_corner/workout_viewer.dart [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+          ],
+        ),
+        Container(
+          color: Colors.white,
+          alignment: Alignment.topCenter,
+          width: MediaQuery.of(context).size.width,
+          height: MediaQuery.of(context).size.height * 0.4,
+          child: AspectRatio(
+              aspectRatio:
+                  16 / 9, // TODO: should be the aspect ratio of the video
```
Comment (lilyydu) [inline lib/widgets/creators_corner/workout_viewer.dart]: Has a ticket been created for this? :) Could we link?

File: lib/widgets/creators_corner/workout_viewer.dart [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+              child: ChewieListItem(
+                  videoPlayerController: VideoPlayerController.network(
+                      'http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4'))),
```
Comment (lilyydu) [inline lib/widgets/creators_corner/workout_viewer.dart]: We should update this too, I'll create & send the mock

**Review by lilyydu — COMMENTED** [review]
Couple of small comments- looks great Lance!!

File: lib/widgets/creators_corner/exercise_library.dart (hunk 6 above)
Comment (lilyydu) [reply lib/widgets/creators_corner/exercise_library.dart]: ^ I didn't include the saved but we can add if its needed, lmk the format if so!

File: lib/widgets/creators_corner/workout_viewer.dart (hunk 7 above)
Comment (lilyydu) [reply lib/widgets/creators_corner/workout_viewer.dart]: There's a couple of comments on Andi's PR so there might be some merging conflicts / fixes you'll need to make

File: lib/widgets/navigation_bar/custom_tab_bar.dart [hunk 9]
```diff
@@ -14,23 +14,70 @@ class CustomTabBar extends StatelessWidget {
...
-    return Padding(
-      padding: EdgeInsets.only(left: screenWidth * 0.05),
-      child: SizedBox(
-        width: screenWidth * tabBarScaling,
-        child: Theme(
-          data: ThemeData(
-            highlightColor: Colors.transparent,
-            splashColor: Colors.transparent,
-            hoverColor: Colors.transparent,
+    return Container(
```
Comment (lilyydu) [inline lib/widgets/navigation_bar/custom_tab_bar.dart:18]: Small design thing but is it possible to make "Creator's Corner" the same font size as "My Workouts" and "Explore"?

File: lib/models/pair.dart (hunk 1 above)
Comment (ltan02) [reply lib/models/pair.dart]: After changing the exercise model, I just forgot to delete this Pair class I made.

File: lib/widgets/navigation_bar/custom_tab_bar.dart (hunk 9 above)
Comment (ltan02) [reply lib/widgets/navigation_bar/custom_tab_bar.dart:18]: We added a ticket for this because it's a scaling issue that I haven't been able to fix.

**Review by richardyjtian — DISMISSED** [review]
Looks good to me! You might need to pull from main and rebase first before being able to merge though

**Review by aleccumming — DISMISSED** [review]
Looks good! Would be great to merge this asap

**Review by lilyydu — DISMISSED** [review]
Thanks for updating Lance! It looks like you have a couple of merge conflicts!

**Review by richardyjtian — APPROVED** [review]

**Review by aleccumming — APPROVED** [review]

**Comment by diogoaprsilva** [comment]
Screenshot 2021-12-08 at 1 26 54 PM

Could not get code to compile, could you please look into this?

**Comment by ltan02** [comment]
> Screenshot 2021-12-08 at 1 26 54 PM
> 
> Could not get code to compile, could you please look into this?

I tried running it on main and also got the same error even though it was working fine on my branch. I tried copying and pasting: 
`intl:
    dependency: "direct main"
    description:
      name: intl
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.17.0"`
It started to work again even though the code is exactly the same. So, not sure why this is happening.

**Comment by aleccumming** [comment]
> > Screenshot 2021-12-08 at 1 26 54 PM
> > Could not get code to compile, could you please look into this?
> 
> I tried running it on main and also got the same error even though it was working fine on my branch. I tried copying and pasting: `intl: dependency: "direct main" description: name: intl url: "https://pub.dartlang.org" source: hosted version: "0.17.0"` It started to work again even though the code is exactly the same. So, not sure why this is happening.

I think you may have forgotten to remove some merge conflict text in pubspec.lock - I see some `<<<<<<< HEAD` business in there.

**Comment by lilyydu** [comment]
Could you try reverting, fixing the code on this PR, and then request to merge again?

Screen Shot 2021-12-08 at 1 46 54 PM

**Comment by diogoaprsilva** [comment]
I fixed the issue with pubspec.yaml now we are missing an asset. 
Screenshot 2021-12-08 at 1 47 38 PM

**Comment by ltan02** [comment]
> I fixed the issue with pubspec.yaml now we are missing an asset. Screenshot 2021-12-08 at 1 47 38 PM

Can you check the assets/images/ folder? On mine, everything is working fine and I don't think the pushup.jpg is actually being used anywhere, which is weird.

**Pull request merged**